# Импорты библиотек

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from datetime import datetime
# Добавьте необходимые библиотеки здесь

# Датасет


**Таблица advertisers** (файл advertisers.csv) В таблице содержится история показов рекламы в мобильных приложениях:

Вот таблица в требуемом формате:

| **Название признака**         | **Описание** |
|-------------------------------|-------------|
| *install*                     | Конверсия показа рекламы в установку приложения (целевая переменная) |
| *user_id*                     | Уникальный идентификатор пользователя, которому была показана реклама |
| *game_app*                    | Идентификатор приложения (игры), в котором была показана реклама |
| *country*                     | Страна пользователя |
| *advertiser*                  | Идентификатор рекламного аукциона, выбравшего рекламу для показа |
| *age*                         | Возрастная группа пользователя (1: 18–25, 2: 26–35, 3: 36–45, 4: 46–55, 5: 55+) |
| *user_quality_score*          | Оценка «качества» пользователя, предсказывающая вероятность конверсии (1 — низкая, 2 — средняя, 3 — высокая) |
| *createdat*                   | Дата и время показа рекламы |
| *spending*                    | Стоимость (расход) на показ рекламы |
| *earning*                     | Плата (доход) от рекламодателя за установку приложения |
| *mccmnc*                      | Код оператора мобильной связи пользователя |


# Чтение данных

In [2]:
advertisers = pd.read_csv('advertisers.csv')

advertisers.head()

install      user id game app country     advertiser  age  \
0        0  user_045828   game_2      AU  advertiser_01    1   
1        0  user_150401   game_4      GB  advertiser_22    5   
2        0  user_102674   game_4      DE  advertiser_07    1   
3        0  user_290419   game_4      AU  advertiser_11    5   
4        0  user_070536   game_5      DE  advertiser_13    4   

   user quality score   createdat  spending  earning     carrier  mccmnc  
0                   1  1675578699  0.014110      0.0  carrier_11  505244  
1                   2  1676420955  0.005462      0.0   carrier_1  234014  
2                   3  1676034552  0.020445      0.0  carrier_17  262203  
3                   3  1677552033  0.010014      0.0   carrier_0  505604  
4                   2  1675775984  0.014186      0.0   carrier_7  262897

In [3]:
print(f"Размер advertisers: {advertisers.shape}") 

Размер advertisers: (50000, 12)


In [4]:
print(f"Типы данных advertisers: {advertisers.dtypes}")  

Типы данных advertisers: install                 int64
user id                object
game app               object
country                object
advertiser             object
age                     int64
user quality score      int64
createdat               int64
spending              float64
earning               float64
carrier                object
mccmnc                  int64
dtype: object


# Чистка данных

In [5]:
print(f"Пропуски в advertisers по колонкам:\n {advertisers.isnull().sum()}")

Пропуски в advertisers по колонкам:
 install               0
user id               0
game app              0
country               0
advertiser            0
age                   0
user quality score    0
createdat             0
spending              0
earning               0
carrier               0
mccmnc                0
dtype: int64


In [6]:
print(f"Дубликаты:\n {advertisers.duplicated().sum()}")

Дубликаты:
 0


In [ ]:
#advertisers.drop_duplicates()

# Анализ

In [7]:
advertisers.describe()

install           age  user quality score     createdat  \
count  50000.000000  50000.000000        50000.000000  5.000000e+04   
mean       0.022040      2.960500            1.983440  1.676420e+09   
std        0.146815      1.425251            0.815806  6.977229e+05   
min        0.000000      1.000000            1.000000  1.675210e+09   
25%        0.000000      2.000000            1.000000  1.675816e+09   
50%        0.000000      3.000000            2.000000  1.676415e+09   
75%        0.000000      4.000000            3.000000  1.677023e+09   
max        1.000000      5.000000            3.000000  1.677629e+09   

           spending       earning         mccmnc  
count  50000.000000  50000.000000   50000.000000  
mean       0.013184      0.013619  382281.645140  
std        0.010679      0.117121  167495.115808  
min        0.000220      0.000000  232014.000000  
25%        0.001081      0.000000  234783.000000  
50%        0.012745      0.000000  312573.000000  
75%        0.020103      0.000000  505231.000000  
max        0.044000      2.000000  724983.000000

In [8]:
advertisers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   install             50000 non-null  int64  
 1   user id             50000 non-null  object 
 2   game app            50000 non-null  object 
 3   country             50000 non-null  object 
 4   advertiser          50000 non-null  object 
 5   age                 50000 non-null  int64  
 6   user quality score  50000 non-null  int64  
 7   createdat           50000 non-null  int64  
 8   spending            50000 non-null  float64
 9   earning             50000 non-null  float64
 10  carrier             50000 non-null  object 
 11  mccmnc              50000 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 4.6+ MB


In [9]:
#Перевод из timestamp в datetime
advertisers['createdat'] = pd.to_datetime(advertisers['createdat'], unit='s')

In [10]:
def age_analysis(df):
    """
    Анализ по возрастным группам
    """
    age_labels = {
        1: '18-25 лет',
        2: '26-35 лет',
        3: '36-45 лет',
        4: '46-55 лет',
        5: '55+ лет'
    }

    age_stats = df['age'].value_counts().sort_index()
    
    print("\nРаспределение показов по возрастам:")
    
    for age, count in age_stats.items():
        label = age_labels.get(age, f'Группа {age}')
        percentage = count / len(df) * 100
        print(f"  {label}: {count:,} ({percentage:.1f}%)")

    print("\nКонверсия по возрастам:")
    conv_stats = df.groupby('age')['install'].mean() * 100
    for age, conv in conv_stats.items():
        label = age_labels.get(age, f'Группа {age}')
        print(f"  {label}: {conv:.2f}%")
        
    best_age = conv_stats.idxmax()
    worst_age = conv_stats.idxmin()
    
    print(f"Лучшая конверсия: {age_labels.get(best_age, best_age)} - {conv_stats[best_age]:.2f}%")
    print(f"Худшая конверсия: {age_labels.get(worst_age, worst_age)} - {conv_stats[worst_age]:.2f}%")
    

    print("\nСредняя стоимость показа по возрастам:")
    spend_stats = df.groupby('age')['spending'].mean()
    for age, spend in spend_stats.items():
        label = age_labels.get(age, f'Группа {age}')
        print(f"  {label}: {spend:.6f}")   
    return age_stats

age_analysis(advertisers)


Распределение по возрастам:
  18-25 лет: 10,701 (21.4%)
  26-35 лет: 9,900 (19.8%)
  36-45 лет: 9,878 (19.8%)
  46-55 лет: 9,715 (19.4%)
  55+ лет: 9,806 (19.6%)

Конверсия по возрастам:
  18-25 лет: 8.31%
  26-35 лет: 1.19%
  36-45 лет: 0.50%
  46-55 лет: 0.36%
  55+ лет: 0.11%
Лучшая конверсия: 18-25 лет - 8.31%
Худшая конверсия: 55+ лет - 0.11%

Средняя стоимость показа по возрастам:
  18-25 лет: $0.013086
  26-35 лет: $0.013239
  36-45 лет: $0.013197
  46-55 лет: $0.013263
  55+ лет: $0.013144


age
1    10701
2     9900
3     9878
4     9715
5     9806
Name: count, dtype: int64

### Вывод по результатам анализа

Критическое неравенство в конверсии
Молодежь 18-25 лет конвертируется в 8.31% случаев - это в 75 раз выше, чем у группы 55+ (0.11%)

Все остальные возрастные группы показывают крайне низкую конверсию (<1.2%)

In [15]:
def conversion_summary(df):
    """
    Вывод конверсии по основным интервалам
    """
    df_temp = df.copy()
    
    df_temp = df_temp[df_temp['age'] == 1]
    
    df_temp['hour'] = df_temp['createdat'].dt.hour
    df_temp['day_of_week'] = df_temp['createdat'].dt.dayofweek
    
    weekdays = ['Пн', 'Вт', 'Ср', 'Чт', 'Пт', 'Сб', 'Вс']
    
    total = len(df_temp)
    installs = df_temp['install'].sum()
    overall_conversion = (installs / total * 100) if total > 0 else 0
    
    print(f"\n ОБЩАЯ СТАТИСТИКА:")
    print(f" Показов: {total:,}")
    print(f" Установок: {installs:,}")
    print(f" Конверсия: {overall_conversion:.2f}%")
    
    hour_conv = df_temp.groupby('hour')['install'].mean() * 100
    
    print(f"\n ТОП-5 ЛУЧШИХ ЧАСОВ:")
    for hour in hour_conv.nlargest(5).index:
        print(f"   {hour:02d}:00 - {hour_conv[hour]:.2f}%")
    
    print(f"\n ТОП-5 ХУДШИХ ЧАСОВ:")
    for hour in hour_conv.nsmallest(5).index:
        print(f"   {hour:02d}:00 - {hour_conv[hour]:.2f}%")
    
    day_conv = df_temp.groupby('day_of_week')['install'].mean() * 100
    
    print(f"\n КОНВЕРСИЯ ПО ДНЯМ:")
    for day in sorted(day_conv.index):
        day_name = weekdays[day] if day < len(weekdays) else f'День {day}'
        print(f"   {day_name}: {day_conv[day]:.2f}%")
    
    best_hour = hour_conv.idxmax()
    worst_hour = hour_conv.idxmin()
    best_day = day_conv.idxmax()
    worst_day = day_conv.idxmin()
    
    best_day_name = weekdays[best_day] if best_day < len(weekdays) else f'День {best_day}'
    worst_day_name = weekdays[worst_day] if worst_day < len(weekdays) else f'День {worst_day}'
    
    print(f"\n ЛУЧШЕЕ ВРЕМЯ:")
    print(f" Час: {best_hour:02d}:00 ({hour_conv[best_hour]:.2f}%)")
    print(f" День: {best_day_name} ({day_conv[best_day]:.2f}%)")
    
    print(f"\n ХУДШЕЕ ВРЕМЯ:")
    print(f" Час: {worst_hour:02d}:00 ({hour_conv[worst_hour]:.2f}%)")
    print(f" День: {worst_day_name} ({day_conv[worst_day]:.2f}%)")

conversion_summary(advertisers)


 ОБЩАЯ СТАТИСТИКА:
 Показов: 10,701
 Установок: 889
 Конверсия: 8.31%

 ТОП-5 ЛУЧШИХ ЧАСОВ:
   20:00 - 11.24%
   14:00 - 11.16%
   03:00 - 11.04%
   23:00 - 9.47%
   22:00 - 9.30%

 ТОП-5 ХУДШИХ ЧАСОВ:
   16:00 - 6.02%
   17:00 - 6.55%
   19:00 - 6.65%
   02:00 - 6.92%
   06:00 - 6.96%

 КОНВЕРСИЯ ПО ДНЯМ:
   Пн: 8.47%
   Вт: 8.25%
   Ср: 7.47%
   Чт: 8.47%
   Пт: 9.44%
   Сб: 7.20%
   Вс: 8.85%

 ЛУЧШЕЕ ВРЕМЯ:
 Час: 20:00 (11.24%)
 День: Пт (9.44%)

 ХУДШЕЕ ВРЕМЯ:
 Час: 16:00 (6.02%)
 День: Сб (7.20%)


### Вывод по результатам анализа конверсии:
Пиковые часы для группы 18-25 лет:

20:00 - 11.24% конверсии

14:00 - 11.16% конверсии

03:00 - 11.04% конверсии

Худшие часы: 16:00 - только 6.02% конверсии

Лучшие дни: Пятница (9.44%), Воскресенье (8.85%)

Худший день: Суббота (7.20%)

In [25]:
def financial_analysis_by_age(df):
    """
    Анализ финансовых показателей по возрастным группам
    """
    age_labels = {
        1: '18-25 лет',
        2: '26-35 лет',
        3: '36-45 лет',
        4: '46-55 лет',
        5: '55+ лет'
    }
    
    results = []
    
    for age in sorted(df['age'].unique()):
        age_df = df[df['age'] == age]
        total_shows = len(age_df)
        
        # Базовые финансовые показатели
        total_spending = age_df['spending'].sum()
        total_earning = age_df['earning'].sum()
        total_installs = age_df['install'].sum()
        
        # Рассчитываем средние значения
        avg_spending_per_show = total_spending / total_shows
        avg_spending_per_install = total_spending / total_installs if total_installs > 0 else 0
        avg_earning_per_show = total_earning / total_shows
        avg_earning_per_install = total_earning / total_installs if total_installs > 0 else 0
        
        # Прибыль и маржа
        total_profit = total_earning - total_spending
        profit_margin = (total_earning - total_spending) / total_spending * 100 if total_spending > 0 else 0
        
        # ROI (Return on Investment)
        roi = (total_earning / total_spending - 1) * 100 if total_spending > 0 else 0
        
        # Конверсия и CPA (Cost Per Acquisition)
        conversion_rate = (total_installs / total_shows) * 100 if total_shows > 0 else 0
        cpa = total_spending / total_installs if total_installs > 0 else 0
        
        # eCPI (effective Cost Per Install) - фактическая стоимость установки
        ecpi = total_spending / total_installs if total_installs > 0 else 0
        
        # LTV:CAC ratio (Lifetime Value to Customer Acquisition Cost)
        # Здесь используем earning как proxy для LTV (в реальности LTV обычно выше)
        ltv_cac_ratio = avg_earning_per_install / cpa if cpa > 0 else 0
        
        results.append({
            'age_group': age_labels.get(age, f'Группа {age}'),
            'age_code': age,
            'total_shows': total_shows,
            'total_installs': total_installs,
            'conversion_rate_%': conversion_rate,
            'total_spending': total_spending,
            'total_earning': total_earning,
            'total_profit': total_profit,
            'profit_margin_%': profit_margin,
            'roi_%': roi,
            'avg_spending_per_show': avg_spending_per_show,
            'avg_earning_per_show': avg_earning_per_show,
            'cpa': cpa,
            'ecpi': ecpi,
            'avg_earning_per_install': avg_earning_per_install,
            'ltv_cac_ratio': ltv_cac_ratio
        })
    
    results_df = pd.DataFrame(results)
    
    print("\n ОБЩАЯ ПРИБЫЛЬНОСТЬ ПО ГРУППАМ:")
    print("-" * 50)
    for _, row in results_df.iterrows():
        profit_sign = "+" if row['total_profit'] >= 0 else ""
        print(f"{row['age_group']}: {profit_sign} {row['total_profit']:.2f} "
              f"(маржа: {row['profit_margin_%']:.1f}%, ROI: {row['roi_%']:.1f}%)")
    
    print("\n ЭФФЕКТИВНОСТЬ КОНВЕРСИИ И ЗАТРАТ:")
    print("-" * 50)
    for _, row in results_df.iterrows():
        print(f"{row['age_group']}: Конверсия: {row['conversion_rate_%']:.2f}%, "
              f"CPA: {row['cpa']:.3f}, eCPI: {row['ecpi']:.3f}")

    print("\n СРЕДНИЕ ПОКАЗАТЕЛИ НА ПОКАЗ:")
    print("-" * 50)
    for _, row in results_df.iterrows():
        print(f"{row['age_group']}: Расход: {row['avg_spending_per_show']:.5f}, "
              f"Доход: {row['avg_earning_per_show']:.5f}")
    
    print("\n АНАЛИЗ РЕНТАБЕЛЬНОСТИ :")
    print("-" * 50)
    sorted_by_profit = results_df.sort_values('total_profit', ascending=False)
    for idx, row in sorted_by_profit.iterrows():
        profit_per_show = row['total_profit'] / row['total_shows']
        print(f"{row['age_group']}: {row['total_profit']:.2f} "
              f"({profit_per_show:.5f}/показ)")

    
    return results_df

financial_analysis_by_age(advertisers)


 ОБЩАЯ ПРИБЫЛЬНОСТЬ ПО ГРУППАМ:
--------------------------------------------------
18-25 лет: + 396.95 (маржа: 283.5%, ROI: 283.5%)
26-35 лет:  -48.35 (маржа: -36.9%, ROI: -36.9%)
36-45 лет:  -94.89 (маржа: -72.8%, ROI: -72.8%)
46-55 лет:  -110.21 (маржа: -85.5%, ROI: -85.5%)
55+ лет:  -121.75 (маржа: -94.5%, ROI: -94.5%)

 ЭФФЕКТИВНОСТЬ КОНВЕРСИИ И ЗАТРАТ:
--------------------------------------------------
18-25 лет: Конверсия: 8.31%, CPA: 0.158, eCPI: 0.158
26-35 лет: Конверсия: 1.19%, CPA: 1.111, eCPI: 1.111
36-45 лет: Конверсия: 0.50%, CPA: 2.660, eCPI: 2.660
46-55 лет: Конверсия: 0.36%, CPA: 3.682, eCPI: 3.682
55+ лет: Конверсия: 0.11%, CPA: 11.718, eCPI: 11.718

 СРЕДНИЕ ПОКАЗАТЕЛИ НА ПОКАЗ:
--------------------------------------------------
18-25 лет: Расход: 0.01309, Доход: 0.05018
26-35 лет: Расход: 0.01324, Доход: 0.00835
36-45 лет: Расход: 0.01320, Доход: 0.00359
46-55 лет: Расход: 0.01326, Доход: 0.00192
55+ лет: Расход: 0.01314, Доход: 0.00073

 АНАЛИЗ РЕНТАБЕЛЬНОСТИ :
--

age_group  age_code  total_shows  total_installs  conversion_rate_%  \
0  18-25 лет         1        10701             889           8.307635   
1  26-35 лет         2         9900             118           1.191919   
2  36-45 лет         3         9878              49           0.496052   
3  46-55 лет         4         9715              35           0.360268   
4    55+ лет         5         9806              11           0.112176   

   total_spending  total_earning  total_profit  profit_margin_%       roi_%  \
0      140.034324     536.980690    396.946366       283.463621  283.463621   
1      131.063365      82.713621    -48.349744       -36.890358  -36.890358   
2      130.355610      35.464138    -94.891472       -72.794314  -72.794314   
3      128.852642      18.647069   -110.205573       -85.528377  -85.528377   
4      128.893604       7.147931   -121.745673       -94.454394  -94.454394   

   avg_spending_per_show  avg_earning_per_show        cpa       ecpi  \
0               0.013086              0.050180   0.157519   0.157519   
1               0.013239              0.008355   1.110706   1.110706   
2               0.013197              0.003590   2.660319   2.660319   
3               0.013263              0.001919   3.681504   3.681504   
4               0.013144              0.000729  11.717600  11.717600   

   avg_earning_per_install  ltv_cac_ratio  
0                 0.604028       3.834636  
1                 0.700963       0.631096  
2                 0.723758       0.272057  
3                 0.532773       0.144716  
4                 0.649812       0.055456

### Выводы по финансовым показателям:
✅ ПРИБЫЛЬНАЯ ГРУППА:

18-25 лет: Прибыль +$396.95 (ROI: +283.5%)

Конверсия: 8.31% (в 7.4 раза выше средней)

CPA: $0.158 (в 9.2 раза дешевле, чем у следующей группы)

Маржа: +283.5% (исключительная рентабельность)

❌ УБЫТОЧНЫЕ ГРУППЫ:

26-35 лет: Убыток -$48.35 (ROI: -36.9%)

36-45 лет: Убыток -$94.89 (ROI: -72.8%)

46-55 лет: Убыток -$110.21 (ROI: -85.5%)

55+ лет: Убыток -$121.75 (ROI: -94.5%)

In [19]:
def quality_analysis(df):
    
    df_temp = df.copy()

    quality_labels = {1: 'Низкое качество', 2: 'Среднее качество', 3: 'Высокое качество'}
    
    conv_by_quality = df_temp.groupby('user quality score')['install'].mean() * 100
    
    print("\n КОНВЕРСИЯ ПО КАЧЕСТВУ:")
    for score in sorted(conv_by_quality.index):
        label = quality_labels.get(score, f'Оценка {score}')
        print(f"  {label}: конверсия {conv_by_quality[score]:.2f}%")
    
    overall_conv = df_temp['install'].mean() * 100
    print(f"\n ОБЩАЯ КОНВЕРСИЯ: {overall_conv:.2f}%")

    return conv_by_quality


quality_analysis(advertisers)


 КОНВЕРСИЯ ПО КАЧЕСТВУ:
  Низкое качество: конверсия 3.97%
  Среднее качество: конверсия 1.89%
  Высокое качество: конверсия 0.67%

 ОБЩАЯ КОНВЕРСИЯ: 2.20%


user quality score
1    3.974442
2    1.885099
3    0.671554
Name: install, dtype: float64

### Вывод по результатам анализа качества пользователей:
Низкое качество (оценка 1): 3.97% конверсии

Среднее качество (оценка 2): 1.89% конверсии

Высокое качество (оценка 3): 0.67% конверсии

Вывод: Оценка "качества" пользователя не работает для прогнозирования конверсии, показывает обратный результат.

# Гипотезы

### Категория 1: Демография и базовые атрибуты пользователя
H1.1: Возраст пользователя является решающим фактором конверсии

H1.2: Страна не влияет на конверсию в возрастной группе 18-25 лет

H1.3: Пользователи с низким user_quality_score конвертируются хуже

H1.4: Страна проживания влияет на конверсию

H1.5: Цена установки (earning) зависит от страны и возраста

H1.6: Мобильный оператор (carrier) влияет на конверсию

H1.7: Операторы в разных странах имеют разную эффективность

H1.8: Некоторые приложения лучше работают с определенными демографическими группами

### Категория 2: Временные паттерны
H2.1: Конверсия зависит от времени суток

H2.2: Выходные дни имеют другую конверсию чем будни

### Категория 3: Источники трафика и платформы (Аукционы и Приложения)
H3.1: Разные аукционы имеют разную эффективность конверсии

H3.2: Эффективные аукционы для молодежи отличаются от эффективных для других возрастов

H3.3: Некоторые аукционы специализируются на определенных странах

H3.4: Разные приложения имеют разную конверсию

H3.5: Конверсия зависит от комбинации приложения и аукциона

### Категория 4: Экономика и прибыльность (Маржа vs Конверсия)
H4.1: Высокая конверсия не значит высокая прибыльность

H4.2: Стоимость показа (spending) коррелирует с вероятностью конверсии

H4.3: Эффект от комбинации факторов сильнее, чем от отдельных факторов (age × country × time)

H4.4: Аукционы с низкой стоимостью показа (spending) имеют лучшую маржу, даже при средней конверсии

H4.5: Прибыль концентрируется в 20% аукционов

H4.6: Лучше покупать много дешевых показов, чем мало дорогих

Н4.7: Аукционы со средней ценой показа и средней конверсией дают лучшую маржу

H4.8: Прибыль можно увеличить на 30+% через исключение убыточных сегментов

H4.9: Оптимальная стратегия: показывать только пользователям 18-25 лет через определенные аукционы

# Вывод
Группа 18-25 лет является ЕДИНСТВЕННОЙ прибыльной возрастной категорией, принося 96% всей прибыли, в то время как все остальные возрастные группы работают в убыток.

# Стратегии проверки ключевых гипотез 
Выделим ключевые гипотезы:

### H1.1: Возраст пользователя является решающим фактором конверсии

    - Данная гипотеза частично подтверждена (конверсия 18-25 лет: 8.31% vs 0.11-1.19% для других групп)

### H4.8: Прибыль можно увеличить на 30+% если исключить возрастные группы 2, 3, 4, 5
    Стратегия проверки:
    
    - Расчет текущей прибыли
    
    - Идентификация убыточных сегментов (возраст 26+)
    
    - Симуляция исключения этих сегментов
    
    - Расчет новой прибыли и % роста

### H1.3: Пользователи с низким user_quality_score конвертируются хуже
    Стратегия проверки:
    
    - Анализ конверсии по user_quality_score
    
    - Сравнение с предсказаниями модели
    
    - Расчет корреляции score с фактической конверсией
    
    - Если гипотеза опровергнута - срочная калибровка модели

### H3.1: Топ-3 аукциона приносят более 70% прибыли для возрастной группы 18-25 лет
    Стратегия проверки:
    
    - Ранжирование всех аукционов по конверсии
    
    - Выявление статистически значимых различий
    
    - ABC-анализ по объему и эффективности
    
    - Исключение аукционов с конверсией ниже порога